File: 942.go
import "slices"

func diStringMatch(s string) []int {
	result := make([]int, len(s)+1)
	for i := range len(result) {
		result[i] = i
	}

	for i := 0; i < len(s); i++ {
		if s[i] == 'I' {
			continue
		}

		j := i + 1
		for j < len(s) && s[j] == 'D' {
			j++
		}

		slices.Reverse(result[i : j+1])
		i = j
	}

	return result
}

File: 942.rs
impl Solution {
    pub fn di_string_match(mut s: String) -> Vec<i32> {
        let mut result: Vec<i32> = (0..(s.len()+1) as i32).collect();

        let mut i = 0;
        while let Some(c) = s.get(i..i+1) {
            if c == "I" {
                i += 1;
                continue;
            }

            let mut j = i + 1;
            while let Some(d) = s.get(j..j+1) {
                if d == "I" {
                    break;
                }
                j += 1;
            }

            j += 1;
            for n in (i..j).rev() {
                result[i] = n as i32;
                i += 1;
            }
        }
        
        result
    }
}
